### PR TITLE
Query linter extension

### LINKED ISSUE #4737: add FROM linter as an extension / option of some kind
Hello, given the following base:

```
class Bar(Base):
    __tablename__ = 'bars'
    id = Column(Integer, primary_key=True)

class Foo(Base):
    __tablename__ = 'foos'

    id = Column(Integer, primary_key=True)

    bar_id = Column(Integer, ForeignKey(Bar.id))
    bar = relationship(Foo, backref='bar')
```

I wonder why sqlalchemy allows this kind of ORM queries to be executed:
```
session.query(Foo).filter(Bar.id == 42)
```
that results in the execution of this SQL:
```
SELECT foos.id AS foos_id, foos.bar_id AS foos_bar_id
FROM foos, bars
WHERE bars.id = ?
```

This is way more often a mistake than what the developer originally intended. I am asking this question because it is quite easy to add a filter condition to a model that was not originally joined, effectively voiding the filter condition:
```
session.query(Foo).join(Foo.bar).filter(Bar.id==42)  # Good query
session.query(Foo).filter(Bar.id==42)  # Oops, I forgot to join Foo.bar.
```
In the above example, the second query will fetch all the `Foo`s from the database. This is both a very slow operation and it also return unwanted records. Maybe it should raise instead?

EDIT: Fixed the last example

### PR DESCRIPTION
**Pull request by youtux**
This is an attempt to make the [FROM linter recipe](https://github.com/sqlalchemy/sqlalchemy/wiki/FromLinter) an extension to SQLAlchemy.
This implementation will just attempt to find unlinked FROM conditions, which lead to cartesian product queries (which is, in general, bad).

### Description
The core logic is completely taken from that recipe. The idea now is to provide:
- a hook that end-users can install on a `before_execute` event on a connection
- a function that, given a Select object, returns the graph of unreachable "from" conditions.

### Checklist
This pull request is:

- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

Fixes #4737.

### CODE REVIEW

**Comment by zzzeek** [comment]
looks good!

**Comment by youtux** [comment]
Hey @zzzeek, how do you want me to continue from here?

**Comment by zzzeek** [comment]
well it needs the pep8 tests to run, "tox -e pep8", so do that it will fail on those long lines, then take it out of draft and I'll run it through CI

**Comment by youtux** [comment]
Fixed the linter issues and marked as ready for review.
Should I rebase on the rel_1_3 branch or should I keep it on master?

**Comment by zzzeek** [comment]
always work with master, I can backport to 1.3

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work to try to get revision f8a21aa6262d1bcc9ff0d11a2616e41fba97a47a of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change f8a21aa6262d1bcc9ff0d11a2616e41fba97a47a: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1461

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

(2 comments)

File: lib/sqlalchemy/ext/linter.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+import collections
```
Comment (sqla-tester) [inline lib/sqlalchemy/ext/linter.py]: **mike bayer** (zzzeek) wrote:

want to try for any docs?  I can do them at some point if not

File: lib/sqlalchemy/ext/linter.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import itertools
+
+from sqlalchemy import util
```
Comment (sqla-tester) [inline lib/sqlalchemy/ext/linter.py]: **mike bayer** (zzzeek) wrote:

these should be relative imports

Comment (youtux) [reply lib/sqlalchemy/ext/linter.py]: Done

File: test/ext/test_linter.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            froms, start = find_unmatching_froms(query, start)
+            assert not froms
+
+    def test_plain_cartesian(self):
+        query = (
+            select([self.a])
+            .where(self.b.c.col_b == 5)
+        )
+        froms, start = find_unmatching_froms(query, self.a)
+        assert start == self.a
```
[the remark this answers is not in the document]
Comment (youtux) [reply test/ext/test_linter.py]: Sure, done.

**Comment by zzzeek** [comment]
I am very happy with this feature but this PR is way down my list of priorities for the moment.    thanks for being patient.

**Comment by auvipy** [comment]
May be good for 1.4/2.0?

**Comment by zzzeek** [comment]
certainly this is useful for 1.4 but I want to consider how it is invoked.   like, could this be a flag on the compiler itself, or an execution option, etc.   im not sure i just want it to be an external thing but I have to look at it more closely.  if I can get it to perform very well maybe it can just be present all the time.   then it needs to be documented once we've decided what it should look like.

**Comment by youtux** [comment]
> I am very happy with this feature but this PR is way down my list of priorities for the moment. thanks for being patient.
I understand that, no problem.

> certainly this is useful for 1.4 but I want to consider how it is invoked. like, could this be a flag on the compiler itself, or an execution option, etc. im not sure i just want it to be an external thing but I have to look at it more closely. if I can get it to perform very well maybe it can just be present all the time. then it needs to be documented once we've decided what it should look like.

Here are my 2 cents on the issue :)

I think this should be treated as an experimental feature for the time being (by not documenting it yet, or but explicitly state it), and just release it in sqlalchemy 1.3/1.4.

Once it is stable enough (not considering performance) then this could be documented properly, invite people to use it, and have a more convenient way to use it (like a query option?).
My reasoning here is that at this point this feature does impact performance, so people would want to enable it on testing/staging environments, but not on live platforms where performance is important.
I currently use this setup with `raiseload` query option: I enable it on the testing environment (for some specific endpoints that must perform well) so that it will fail in case of N+1 queries, but on the real platform I disable it, so that it can gracefully degrade to suboptimal performance without breaking the functionality.

If it will ever be so performant that it can always be enabled, it could be enabled enabled more easily, possibly by a flag to the engine/session factory.

And maybe default this flag to True in a major sqlalchemy release (sqlalchemy 2.0?).

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

I'm still really curious if i can get this to just be built-in.    there are a lot of moves I'm working on for 1.4 and beyond that can make this even more feasible (more statement caching).

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

I've repurposed this as a built in flag in https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1664/ .   it has been invaluable to be present by default for the full test suite and initial responses from twitter users has been very positive.   thanks for the help on this!

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/1461 has been **abandoned**.  That means that at least for the moment I need to close this pull request. Sorry it didn't work out :(

**Comment by youtux** [comment]
Nice, I didn't expect this to become a core feature! I'm glad to hear that it helped to improve the project!

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Verified+1 Code-Review+2 Workflow+1

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/1664 has been **merged**. Congratulations! :)

**Pull request closed without merging**